src/networking.rs

```diff
@@ -0,0 +1,191 @@
+use subspace_farmer::single_disk_plot::SingleDiskPlot;
+use subspace_networking::{LimitedSizeRecordStorageWrapper, ParityDbRecordStorage};
+use subspace_service::piece_cache::PieceCache;
+
+#[derive(Debug, Copy, Clone)]
+pub struct PieceDetails {
+    pub plot_offset: usize,
+    pub sector_index: SectorIndex,
+    pub piece_offset: u64,
+}
+
+#[derive(Debug)]
+pub struct ReadersAndPieces {
+    readers: Vec<PieceReader>,
+    pieces: std::collections::HashMap<PieceIndexHash, PieceDetails>,
+    handle: tokio::runtime::Handle,
+}
+
+impl ReadersAndPieces {
+    pub async fn new(single_disk_plots: &[SingleDiskPlot]) -> Self {
+        // Store piece readers so we can reference them later
+        let readers = single_disk_plots.iter().map(SingleDiskPlot::piece_reader).collect();
+
+        tracing::debug!("Collecting already plotted pieces");
+
+        // Collect already plotted pieces
+        let pieces = single_disk_plots
+            .iter()
+            .enumerate()
+            .flat_map(|(plot_offset, single_disk_plot)| {
+                single_disk_plot
+                    .plotted_sectors()
+                    .enumerate()
+                    .filter_map(move |(sector_offset, plotted_sector_result)| {
+                        match plotted_sector_result {
+                            Ok(plotted_sector) => Some(plotted_sector),
+                            Err(error) => {
+                                tracing::error!(
+                                    %error,
+                                    %plot_offset,
+                                    %sector_offset,
+                                    "Failed reading plotted sector on startup, skipping"
+                                );
+                                None
+                            }
+                        }
+                    })
+                    .flat_map(move |plotted_sector| {
+                        plotted_sector.piece_indexes.into_iter().enumerate().map(
+                            move |(piece_offset, piece_index)| {
+                                (
+                                    PieceIndexHash::from_index(piece_index),
+                                    PieceDetails {
+                                        plot_offset,
+                                        sector_index: plotted_sector.sector_index,
+                                        piece_offset: piece_offset as u64,
+                                    },
+                                )
+                            },
+                        )
+                    })
+            })
+            // We implicitly ignore duplicates here, reading just from one of the plots
+            .collect();
+
+        tracing::debug!("Finished collecting already plotted pieces");
+
+        let handle = tokio::runtime::Handle::current();
+        Self { readers, pieces, handle }
+    }
+
+    pub fn get_piece(&self, key: &PieceIndexHash) -> Option<Option<Piece>> {
```

Comment:
I'm sure it has its reasons, but `Option<Option<Piece>>` seems like a weird API. Are you sure this is the best way?